lib/ansible/modules/cloud/amazon/iam_server_certificate_facts.py

```diff
@@ -81,89 +89,88 @@
+    """
+    Get one or more IAM server certificates facts based on name or path. If not found, return None.
 
-def get_server_certs(iam, name=None):
-    """Retrieve the attributes of a server certificate if it exists or all certs.
-    Args:
-        iam (botocore.client.IAM): The boto3 iam instance.
-
-    Kwargs:
-        name (str): The name of the server certificate.
-
-    Basic Usage:
-        >>> import boto3
-        >>> iam = boto3.client('iam')
-        >>> name = "server-cert-name"
-        >>> results = get_server_certs(iam, name)
-        {
-            "upload_date": "2015-04-25T00:36:40+00:00",
-            "server_certificate_id": "ADWAJXWTZAXIPIMQHMJPO",
-            "certificate_body": "-----BEGIN CERTIFICATE-----\nbunch of random data\n-----END CERTIFICATE-----",
-            "server_certificate_name": "server-cert-name",
-            "expiration": "2017-06-15T12:00:00+00:00",
-            "path": "/",
-            "arn": "arn:aws:iam::911277865346:server-certificate/server-cert-name"
-        }
+    :param connection: AWS boto3 glue connection
+    :param module: Ansible module
+    :param name: The name of the server certificate
+    :param path: The path of the server certificate
+    :return: IAM server certificates dict or None if not found
     """
-    results = dict()
+
+    response = dict()
+    response['certificates'] = []
+
     try:
         if name:
-            server_certs = [iam.get_server_certificate(ServerCertificateName=name)['ServerCertificate']]
+            response['certificates'].append(connection.get_server_certificate(ServerCertificateName=name))
+        elif path:
+            server_certs = connection.list_server_certificates(PathPrefix=path)['ServerCertificateMetadataList']
         else:
-            server_certs = iam.list_server_certificates()['ServerCertificateMetadataList']
+            server_certs = connection.list_server_certificates()['ServerCertificateMetadataList']
 
-        for server_cert in server_certs:
-            if not name:
-                server_cert = iam.get_server_certificate(ServerCertificateName=server_cert['ServerCertificateName'])['ServerCertificate']
-            cert_md = server_cert['ServerCertificateMetadata']
-            results[cert_md['ServerCertificateName']] = {
-                'certificate_body': server_cert['CertificateBody'],
-                'server_certificate_id': cert_md['ServerCertificateId'],
-                'server_certificate_name': cert_md['ServerCertificateName'],
-                'arn': cert_md['Arn'],
-                'path': cert_md['Path'],
-                'expiration': cert_md['Expiration'].isoformat(),
-                'upload_date': cert_md['UploadDate'].isoformat(),
-            }
+        if not name:
+            for server_cert in server_certs:
+                response['certificates'].append(connection.get_server_certificate(ServerCertificateName=server_cert['ServerCertificateName']))
```

Comment:
This actually breaks the existing return format. The previous code creates a dictionary like:

```
{
    'MyCertName': { .... cert info object ...},
    'AnotherCertName': { .... cert info object ...},
}
```

The new code is instead creating a list, like:

```
[
     { .... cert info object ...},
     { .... cert info object ...},
]
```

We can't break backwards compatibility on the return of these facts.

Reply:
@ryansb the module is marked as 'preview' so could we look at making this change?  It brings it in to line with other aws _facts modules by:

 - returning a list of dicts, with the cert name included in that dict
 - returning an empty list if not found rather than failing

Every other _facts module i can think of behaves this way.  I know it's a breaking change but i guess that's what the 'preview' type is for right?

Reply:
@ryansb I agree with @wimnat. The existing module is not inlines with the other _facts module. 

Also please note that the existing module although contains the code to retrieve all server certificates, it is not used since `name` is marked as required. I have now added both `name` and `path` attribute to make this module more complete. 

With regards to the module output, I just tried to make it consistent with the other _facts modules.

Reply:
I can see the value in it being a dictionary though, so you cloud do something like `cert_facts.certificates['www.foo.com']` instead of needing to search through the list. Is there a use case where the list is more useful? I think `cert_facts.certificates.values()` would be an ok way to get the list of certs

Reply:
@ryansb, the use case here is to sort certs based on the upload date and use the most recent certificate when creating ELB's or ALB's. 

Assuming we need a certificate with a particular name, we can use the module with the `name` option which will return one list object that matches the `name`.

Another important reason for returning a list instead of the dict is to make this module consistent with the other _facts module which return a list of objects. For example: 

ec2_vpc_net_facts returns: 

`{ 
        "vpcs": [ ......
        ]
}`

ec2_vpc_subnet_facts returns: 

`{ 
        "subnets": [ ......
        ]
}`

Reply:
It's true that it's useful to have a dict with the name as a key *if* that's what you're interested in.  But who's to say name should be the key?  Why not path, or certificate id?

With a list you can choose whatever you want using jinja selectattr.

E.g.

`certs.certificates | selectattr('server_certificate_metadata.server_certificate_name', 'equalto', 'www.foo.com') | list`
or if you wanted id
`certs.certificates | selectattr('server_certificate_metadata.server_certificate_id', 'equalto', 'xyz') | list`

This is how i select subnets in a particular AZ from the vpc subnet facts module.  I wouldn't expect a key of AZ because there's a possibility someone might want vpc_id or something else.

As for a use case for this module, I like this change because it would allow me to easily pick the cert that has the latest expiry date.  Something like:

```
"{{ certs.certificates | selectattr('server_certificate_metadata.path', 'equalto', '/') | sort(attribute='server_certificate_metadata.expiration') | last }}"
```

Reply:
Hi @ryansb, 

Did you manage have look at my comments? Appreciate your feedback on this.